File: src/backend/lib/room.py
from dataclasses import dataclass, field
from typing import Optional

from .board import Board

@dataclass
class Player:
    username: str
    progress: int = 0

@dataclass
class Room:
    players: list[Player]
    board: Board
    seed: str
    has_started: bool = False
    next_room: Optional[int] = None
    letter_count: int = 7
    minimum_word_length: int = 3
    ghost_players: list[str] = field(default_factory=list) # Players that have left the room but are still stored for the leaderboard

    def to_json(self, room_id, username):
        return {
            "roomId": room_id,
            "username": username,
            "width": self.board.width,
            "height": self.board.height,
            "hasStarted": self.has_started,
            "words": [{
                "x": word[0],
                "y": word[1],
                "direction": word[2].value,
                "word": word[3]
            } for word in self.board.words],
            "players": [{
                "username": player.username,
                "progress": player.progress
            } for player in self.players],
            "letters": self.board.letters,
            "seed": self.seed,
            "letterCount": self.letter_count,
            "minimumWordLength": self.minimum_word_length
        }

File: src/frontend/src/lib/room.ts
import { room_store } from "./stores";

export type Word = {
    x: number,
    y: number,
    direction: string,
    word: string,
}

export type Player = {
    username: string,
    progress: number,
}

export type Room = {
    username: string,
    players: Player[],
    roomId: number,
    words: Word[],
    width: number,
    height: number,
    letters: string[],
    seed: string,
    hasStarted: boolean,
    letterCount: number,
    minimumWordLength: number,
    winner?: string,
};

export const shuffleLetters = () => {
    room_store.update(room => {
        while (true) {
            let currentIndex = room.letters.length, randomIndex;

            while (currentIndex != 0) {
                randomIndex = Math.floor(Math.random() * currentIndex);
                currentIndex--;

                [room.letters[currentIndex], room.letters[randomIndex]] = [
                room.letters[randomIndex], room.letters[currentIndex]];
            }

            // Make sure that we don't show them the 7 letter word.
            let isWord = false;
            let shownWord = room.letters.join();
            for (const word of room.words) {
                if (shownWord == word.word) {
                    isWord = true;
                    break
                }
            }

            if (!isWord) {
                break
            }
        }

        return room
    })
}